pygfx/renderers/wgpu/_environment.py

```diff
@@ -0,0 +1,167 @@
+
+        if renderer in renderers:
+            if renderer_state_hash != self._renderer_state_hash:
+                self._renderers.discard(renderer)
+                renderers.discard(renderer)
+
+        if scene in scenes:
+            if scene_state_hash != self._scene_state_hash:
+                self._scenes.discard(scene)
+                scenes.discard(renderer)
+
+        if not renderers or not scenes:
+            self.clear()
+            return True
+
+    def clear(self):
+        """Remove all wgpu objects associated with this environment."""
+        for pipeline_container in self._pipeline_containers:
+            pipeline_container.remove_env_hash(self.hash)
+        self._pipeline_containers.clear()
+        self._renderers.clear()
+        self._scenes.clear()
+
+
+class GlobalEnvironmentManager:
+    """A little class to manage the different environments."""
+
+    def __init__(self):
+        self.environments = {}  # hash -> Environment
+
+    def get_environment(self, renderer, scene):
+        """The main entrypoint. The renderer uses this to obtain an
+        environment object.
+        """
+        renderer_state_hash, scene_state_hash, state = get_hash_and_state(
+            renderer, scene
+        )
+        env_hash = (renderer_state_hash, scene_state_hash)
+
+        # Re-use or create an environment
+        if env_hash in self.environments:
+            env = self.environments[env_hash]
+        else:
+            env = Environment(renderer_state_hash, scene_state_hash)
+            assert env.hash == env_hash
+            self.environments[env_hash] = env
+
+        # Update the environment
+        env.update(renderer, scene, **state)
+
+        # Cleanup
+        self._cleanup(renderer, scene, renderer_state_hash, scene_state_hash)
+
+        return env
+
+    def _cleanup(self, renderer, scene, renderer_state_hash, scene_state_hash):
+        """ " Remove all environments of which the associated renderer
+        or scene no longer exists, or their states have changed.
+        """
+        hashes_to_drop = []
+        for env_hash, env in self.environments.items():
+            if env.check_inactive(
+                renderer, scene, renderer_state_hash, scene_state_hash
+            ):
+                hashes_to_drop.append(env_hash)
+        for env_hash in hashes_to_drop:
+            self.environments.pop(env_hash)
+
+
+environment_manager = GlobalEnvironmentManager()
+get_environment = environment_manager.get_environment
+
+
+def get_hash_and_state(renderer, scene):
+
+    state = {}
+
+    # For renderer
+
+    state["blender"] = renderer._blender
+    renderer_hash = renderer.blend_mode
+
+    # For scene
+
+    lights = []
+
+    def visit(ob):
+        pass
+        # if isinstance(ob, Light):
+        #     lights.append(ob)
+
+    scene.traverse(visit)
+
+    state["lights"] = lights
+    scene_hash = (len(lights),)
+
+    return renderer_hash, scene_hash, state
```

Comment:
is this a temporary function?

Reply:
This code is ready to include scene-related things like lights, but does not actually do anything. I will comment out some lines here.